File: hash_table/javascript/hash_table_test.js
import assert from 'assert';
import HashTable from './hash_table';

function testBasicSequence() {
    const h = new HashTable(10);

    assert.ok(!h.has('abc'));

    h.set('abc', 5);
    assert.ok(h.has('abc'));
    assert.strictEqual(h.get('abc'), 5);

    // Check that deleting a non-existing key doesn't crash.
    h.del('aaa');
    h.del('zzz');

    h.set('abc', 7);
    assert.ok(h.has('abc'));
    assert.strictEqual(h.get('abc'), 7);

    h.set('def', 9);
    assert.ok(h.has('abc'));
    assert.strictEqual(h.get('abc'), 7);
    assert.ok(h.has('def'));
    assert.strictEqual(h.get('def'), 9);

    h.del('abc');
    assert.ok(!h.has('abc'));
    assert.ok(h.has('def'));
    assert.strictEqual(h.get('def'), 9);
}

function testRandomSequences() {
    for (let i = 0; i < 100; i += 1) {
        testRandomSequence();
    }
}

function testRandomSequence() {
    const h = new HashTable(100);
    const m = new Map();
    const a = [];
    const count = Math.floor(Math.random() * 1000);

    for (let i = 0; i < count; i += 1) {
        randomOperation(h, m, a);
    }

    validate(h, m, a);
}

function validate(h, m, a) {
    for (const k of a) {
        assert.ok(h.has(k));
        assert.strictEqual(m.get(k), h.get(k));
    }
}

function randomOperation(h, m, a) {
    const r = Math.random();
    if (r < 0.2) {
        delRandom(h, m, a);
    } else {
        setRandom(h, m, a);
    }
}

function setRandom(h, m, a) {
    const k = randomKey();
    const v = randomValue();
    h.set(k, v);
    if (!m.has(k)) {
        a.push(k);
    }
    m.set(k, v);
}

function delRandom(h, m, a) {
    if (!a.length) {
        return;
    }
    const i = Math.floor(Math.random() * a.length);
    const k = a[i];
    h.del(k);
    m.delete(k);
    a.splice(i, 1);
}

function randomKey() {
    // This random string generator is not uniform at all, but it doesn't matter for this test.
    const length = Math.floor(1 + Math.random() * 10);
    return (Math.PI + Math.random()).toString(36).substr(2, length);
}

function randomValue() {
    return Math.floor(Math.random() * 1e9);
}

testBasicSequence();
testRandomSequences();

File: hash_table/go/hash_table_test.go
package hash_table

import (
	"math/rand"
	"testing"
)

var chars = []byte("01234567890abcdefghijklmnopqrstuvwxyz")

func TestBasicSequence(t *testing.T) {
	h := NewHashTable(10)

	if _, ok := h.Get("abc"); ok {
		t.Error("Expected ok to be false, was true")
	}

	h.Set("abc", 5)
	if v, ok := h.Get("abc"); !ok || v != 5 {
		t.Errorf("Expected (ok, v) to be (true, 5), was (%t, %d)", ok, v)
	}

	// Check that deleting a non-existing key doesn't crash.
	h.Del("aaa")
	h.Del("zzz")

	h.Set("abc", 7)
	if v, ok := h.Get("abc"); !ok || v != 7 {
		t.Errorf("Expected (ok, v) to be (true, 7), was (%t, %d)", ok, v)
	}

	h.Set("def", 9)
	if v, ok := h.Get("abc"); !ok || v != 7 {
		t.Errorf("Expected (ok, v) to be (true, 7), was (%t, %d)", ok, v)
	}
	if v, ok := h.Get("def"); !ok || v != 9 {
		t.Errorf("Expected (ok, v) to be (true, 9), was (%t, %d)", ok, v)
	}

	h.Del("abc")
	if _, ok := h.Get("abc"); ok {
		t.Error("Expected ok to be false, was true")
	}
	if v, ok := h.Get("def"); !ok || v != 9 {
		t.Errorf("Expected (ok, v) to be (true, 9), was (%t, %d)", ok, v)
	}
}

func TestRandomSequences(t *testing.T) {
	for i := 0; i < 100 && !t.Failed(); i++ {
		testRandomSequence(t)
	}
}

func testRandomSequence(t *testing.T) {
	h := NewHashTable(100)
	m := map[string]int{}
	a := []string{}

	count := rand.Intn(1000)
	for i := 0; i < count; i++ {
		randomOperation(h, m, &a)
		if rand.Float32() < 0.2 {
			delRandom(h, m, &a)
			continue
		}
		setRandom(h, m, &a)
	}

	validate(t, h, m, a)
}

func validate(t *testing.T, h *HashTable, m map[string]int, a []string) {
	for _, k := range a {
		if v, ok := h.Get(k); !ok || v != m[k] {
			t.Errorf("On Get(%q), expected (ok, v) to be (true, %d), was (%t, %d)", k, m[k], ok, v)
		}
	}
}

func randomOperation(h *HashTable, m map[string]int, a *[]string) {
	r := rand.Float32()
	switch {
	case r < 0.2:
		delRandom(h, m, a)
	default:
		setRandom(h, m, a)
	}
}

func setRandom(h *HashTable, m map[string]int, a *[]string) {
	k := randomKey()
	v := randomValue()
	h.Set(k, v)
	if _, ok := m[k]; !ok {
		(*a) = append(*a, k)
	}
	m[k] = v
}

func delRandom(h *HashTable, m map[string]int, a *[]string) {
	l := len(*a)
	if l <= 0 {
		return
	}
	i := rand.Intn(l)
	k := (*a)[i]
	h.Del(k)
	delete(m, k)
	(*a)[i] = (*a)[l-1]
	(*a) = (*a)[:l-1]
}

func randomKey() string {
	l := 1 + rand.Intn(10)
	c := len(chars)
	s := []byte{}
	for i := 0; i < l; i++ {
		s = append(s, chars[rand.Intn(c)])
	}
	return string(s)
}

func randomValue() int {
	return rand.Int()
}
